src/Simd/MatrixFloat2x2.cs

```diff
@@ -0,0 +1,146 @@
+//
+// Authors:
+//     Rolf Bjarne Kvinge <rolf@xamarin.com>
+//
+// Copyright (c) 2017 Microsoft Inc
+//
+
+// 
+// This represents the native matrix_float2x2 type, which has a column-major layout
+// (as opposed to OpenTK.Matrix2, which has a row-major layout).
+// 
+
+using System;
+using System.Runtime.InteropServices;
+
+using VectorFloat2=global::OpenTK.Vector2;
+
+namespace Simd
+{
+	[StructLayout (LayoutKind.Sequential)]
+	public struct MatrixFloat2x2 : IEquatable<MatrixFloat2x2>
+	{
+		public float M11;
+		public float M21;
+		public float M12;
+		public float M22;
+
+		public readonly static MatrixFloat2x2 Identity = new MatrixFloat2x2 (
+			1, 0,
+			0, 1);
+
+		public MatrixFloat2x2 (VectorFloat2 column0, VectorFloat2 column1)
+		{
+			M11 = column0.X;
+			M21 = column0.Y;
+			M12 = column1.X;
+			M22 = column1.Y;
+		}
+
+		public MatrixFloat2x2 (
+			float m11, float m12,
+			float m21, float m22)
+		{
+			M11 = m11;
+			M21 = m21;
+			M12 = m12;
+			M22 = m22;
+		}
+
+		public float Determinant {
+			get {
+				return M11 * M22 - M21 * M12;
+			}
+		}
+
+		public void Transpose ()
+		{
+			this = Transpose (this);
+		}
+
+		public static MatrixFloat2x2 Transpose (MatrixFloat2x2 mat)
+		{
+			return new MatrixFloat2x2 (mat.M11, mat.M21, mat.M12, mat.M22);
+		}
+
+		public static void Transpose (ref MatrixFloat2x2 mat, out MatrixFloat2x2 result)
+		{
+			result.M11 = mat.M11;
+			result.M12 = mat.M21;
+			result.M21 = mat.M12;
+			result.M22 = mat.M22;
+		}
+
+		public static MatrixFloat2x2 Multiply (MatrixFloat2x2 left, MatrixFloat2x2 right)
+		{
+			MatrixFloat2x2 result;
+			Multiply (ref left, ref right, out result);
+			return result;
+		}
+
+		public static void Multiply (ref MatrixFloat2x2 left, ref MatrixFloat2x2 right, out MatrixFloat2x2 result)
+		{
+			result.M11 = left.M11 * right.M11 + left.M12 * right.M21;
+			result.M12 = left.M11 * right.M12 + left.M12 * right.M22;
+
+			result.M21 = left.M21 * right.M11 + left.M22 * right.M21;
+			result.M22 = left.M21 * right.M12 + left.M22 * right.M22;
+		}
+
+		public static MatrixFloat2x2 operator * (MatrixFloat2x2 left, MatrixFloat2x2 right)
+		{
+			return Multiply (left, right);
+		}
+
+		public static bool operator == (MatrixFloat2x2 left, MatrixFloat2x2 right)
+		{
+			return left.Equals (right);
+		}
+
+		public static bool operator != (MatrixFloat2x2 left, MatrixFloat2x2 right)
+		{
+			return !left.Equals (right);
+		}
+
+		public static explicit operator global::OpenTK.Matrix2 (MatrixFloat2x2 value)
```

Comment:
Question: should this be an implicit operator instead? The conversion is lossless (it's just transposing the matrix), but it could also lead to slow code (it would be very easy to write code that does a lot of conversions, which would slow it down).

Reply:
No, it's too easy to hide the performance cost. The overuse of the explicit one should be spotted in code reviews as a potential "non-free" issue.